File: baml_language/crates/bex_query/src/bqf1.rs
//! §9.3 BQF1: fixed little-endian columnar frames.
//!
//! Not Arrow (schema machinery buys nothing for ~10 known kinds), not JSON
//! (the measured 2.21 GB). Layout:
//!
//! ```text
//! [40 B header]
//!   magic        4B  "BQF1"
//!   kind         u16          (FrameKind)
//!   flags        u16          (bit0 lod_degraded, bit1 partial_tail, bit2 more_lanes)
//!   request_id   u64
//!   epoch        u64          (data generation the frame was computed at)
//!   ncols        u32
//!   nrows        u32
//!   payload_len  u64          (directory + columns, excluding trailer)
//! [column directory]  ncols × 16 B: (col_type u8, pad [u8;7], offset u64 from payload start)
//! [column payloads]   each 8-byte aligned
//!   fixed cols:  nrows × elem_size
//!   str cols:    (nrows+1) × u32 offsets, pad8, then utf8 bytes
//! [8 B trailer]  crc32c(header + directory + payloads) as u64 LE
//! ```
//!
//! Decodes into zero-copy `TypedArray` views in ~150 lines of TS.

use bex_events::prof::cct::crc32c;

pub const BQF1_MAGIC: [u8; 4] = *b"BQF1";
pub const HEADER_LEN: usize = 40;

pub const FLAG_LOD_DEGRADED: u16 = 1 << 0;
pub const FLAG_PARTIAL_TAIL: u16 = 1 << 1;
pub const FLAG_MORE_LANES: u16 = 1 << 2;

/// Frame kinds — the ~10 known shapes of §9.3.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u16)]
pub enum FrameKind {
    /// One row per run: the §9.6 runs list.
    RunsList = 1,
    /// Run header + the function dictionary (id ↔ fqn) sent once.
    RunMeta = 2,
    /// Activity bands per (thread × window bucket) — §9.4 aggregate tier.
    Timeline = 3,
    /// Preorder Left-Heavy fold rows.
    LeftHeavy = 4,
    /// Per-function totals table.
    TopFunctions = 5,
    /// Error/status report for a request.
    Status = 6,
    /// Live totals tick (small; live-tail subscriptions).
    LiveTotals = 7,
    /// §9.4 exact-recency tier: recent-call rects.
    RecentCalls = 8,
    /// §8 BQL result: a generic typed table with free-form columns.
    ///
    /// Layout convention (documented here because the columns are not a
    /// fixed shape like the other kinds):
    /// - The LAST column is always a `Str` meta column. Its row 0 carries
    ///   one JSON object `{"columns":[{"name","type"},...],"rows":N,
    ///   "footer":{"sealed","torn","first_ts_ns","last_ts_ns",
    ///   "degraded":[...]}}` — the §8.4 completeness footer every BQL
    ///   result must ship. Rows 1.. of the meta column are empty strings.
    /// - The preceding columns are the data columns named by the meta
    ///   JSON. Frame row 0 is a sentinel row (numeric 0 / empty string);
    ///   data row `i` lives at frame row `i + 1`.
    /// - `nrows = data_rows + 1`, so an EMPTY result is a 1-row frame that
    ///   still carries its footer.
    BqlTable = 9,
}

/// Column element types (u8 in the directory).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum ColType {
    U32 = 1,
    U64 = 2,
    F64 = 3,
    /// (nrows+1) u32 offsets + utf8 bytes.
    Str = 4,
}

/// One column of data to encode.
pub enum Col<'a> {
    U32(&'a [u32]),
    U64(&'a [u64]),
    F64(&'a [f64]),
    Str(&'a [String]),
}

impl Col<'_> {
    fn col_type(&self) -> ColType {
        match self {
            Col::U32(_) => ColType::U32,
            Col::U64(_) => ColType::U64,
            Col::F64(_) => ColType::F64,
            Col::Str(_) => ColType::Str,
        }
    }

    fn nrows(&self) -> usize {
        match self {
            Col::U32(v) => v.len(),
            Col::U64(v) => v.len(),
            Col::F64(v) => v.len(),
            Col::Str(v) => v.len(),
        }
    }
}

fn pad8(out: &mut Vec<u8>) {
    while out.len() % 8 != 0 {
        out.push(0);
    }
}

/// Encode one frame. All columns must share `nrows` (asserted).
#[must_use]
pub fn encode_frame(
    kind: FrameKind,
    flags: u16,
    request_id: u64,
    epoch: u64,
    cols: &[Col<'_>],
) -> Vec<u8> {
    let nrows = cols.first().map_or(0, Col::nrows);
    for c in cols {
        assert_eq!(c.nrows(), nrows, "BQF1 columns must share nrows");
    }

    // Payloads, collecting per-column offsets relative to payload start
    // (payload start = after the directory).
    let mut payload = Vec::new();
    let mut offsets = Vec::with_capacity(cols.len());
    for col in cols {
        pad8(&mut payload);
        offsets.push(payload.len() as u64);
        match col {
            Col::U32(v) => {
                for x in *v {
                    payload.extend_from_slice(&x.to_le_bytes());
                }
            }
            Col::U64(v) => {
                for x in *v {
                    payload.extend_from_slice(&x.to_le_bytes());
                }
            }
            Col::F64(v) => {
                for x in *v {
                    payload.extend_from_slice(&x.to_le_bytes());
                }
            }
            Col::Str(v) => {
                let mut at: u32 = 0;
                payload.extend_from_slice(&at.to_le_bytes());
                let mut bytes = Vec::new();
                for s in *v {
                    at = at.saturating_add(u32::try_from(s.len()).unwrap_or(u32::MAX));
                    bytes.extend_from_slice(s.as_bytes());
                    payload.extend_from_slice(&at.to_le_bytes());
                }
                pad8(&mut payload);
                payload.extend_from_slice(&bytes);
            }
        }
    }
    pad8(&mut payload);

    let dir_len = cols.len() * 16;
    let payload_len = (dir_len + payload.len()) as u64;

    let mut out = Vec::with_capacity(HEADER_LEN + dir_len + payload.len() + 8);
    out.extend_from_slice(&BQF1_MAGIC);
    out.extend_from_slice(&(kind as u16).to_le_bytes());
    out.extend_from_slice(&flags.to_le_bytes());
    out.extend_from_slice(&request_id.to_le_bytes());
    out.extend_from_slice(&epoch.to_le_bytes());
    out.extend_from_slice(&u32::try_from(cols.len()).unwrap_or(u32::MAX).to_le_bytes());
    out.extend_from_slice(&u32::try_from(nrows).unwrap_or(u32::MAX).to_le_bytes());
    out.extend_from_slice(&payload_len.to_le_bytes());
    debug_assert_eq!(out.len(), HEADER_LEN);

    for (col, off) in cols.iter().zip(&offsets) {
        out.push(col.col_type() as u8);
        out.extend_from_slice(&[0u8; 7]);
        out.extend_from_slice(&off.to_le_bytes());
    }
    out.extend_from_slice(&payload);

    let crc = u64::from(crc32c::crc32c(&out));
    out.extend_from_slice(&crc.to_le_bytes());
    out
}

/// A convenience Status frame: one row, `code` + `message` columns.
#[must_use]
pub fn status_frame(request_id: u64, code: u32, message: &str) -> Vec<u8> {
    encode_frame(
        FrameKind::Status,
        0,
        request_id,
        0,
        &[
            Col::U32(&[code]),
            Col::Str(std::slice::from_ref(&message.to_string())),
        ],
    )
}

/// Decoded frame view (host-side tests; the TS decoder is the real
/// consumer). Owns nothing — borrows the frame bytes.
#[derive(Debug)]
pub struct FrameView<'a> {
    pub kind: u16,
    pub flags: u16,
    pub request_id: u64,
    pub epoch: u64,
    pub nrows: u32,
    pub cols: Vec<(ColType, &'a [u8])>,
    payload: &'a [u8],
}

impl<'a> FrameView<'a> {
    pub fn col_u32(&self, i: usize) -> Option<Vec<u32>> {
        let (ColType::U32, bytes) = self.cols.get(i)? else {
            return None;
        };
        Some(
            bytes
                .chunks_exact(4)
                .take(self.nrows as usize)
                .map(|c| u32::from_le_bytes(c.try_into().unwrap()))
                .collect(),
        )
    }

    pub fn col_u64(&self, i: usize) -> Option<Vec<u64>> {
        let (ColType::U64, bytes) = self.cols.get(i)? else {
            return None;
        };
        Some(
            bytes
                .chunks_exact(8)
                .take(self.nrows as usize)
                .map(|c| u64::from_le_bytes(c.try_into().unwrap()))
                .collect(),
        )
    }

    pub fn col_str(&self, i: usize) -> Option<Vec<String>> {
        let (ColType::Str, bytes) = self.cols.get(i)? else {
            return None;
        };
        let n = self.nrows as usize;
        let mut offsets = Vec::with_capacity(n + 1);
        for c in bytes.chunks_exact(4).take(n + 1) {
            offsets.push(u32::from_le_bytes(c.try_into().unwrap()) as usize);
        }
        // The utf8 blob starts at the 8-aligned boundary after offsets.
        let mut blob_at = (n + 1) * 4;
        blob_at += (8 - blob_at % 8) % 8;
        let blob = &bytes[blob_at..];
        let mut out = Vec::with_capacity(n);
        for w in offsets.windows(2) {
            out.push(String::from_utf8_lossy(blob.get(w[0]..w[1])?).into_owned());
        }
        Some(out)
    }

    /// The payload region (for size accounting in tests).
    #[must_use]
    pub fn payload_len(&self) -> usize {
        self.payload.len()
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum DecodeError {
    TooShort,
    BadMagic,
    BadCrc,
    BadDirectory,
}

/// Host-side decode (tests + native round-trips).
pub fn decode_frame(bytes: &[u8]) -> Result<FrameView<'_>, DecodeError> {
    if bytes.len() < HEADER_LEN + 8 {
        return Err(DecodeError::TooShort);
    }
    if bytes[0..4] != BQF1_MAGIC {
        return Err(DecodeError::BadMagic);
    }
    let body = &bytes[..bytes.len() - 8];
    let crc = u64::from_le_bytes(bytes[bytes.len() - 8..].try_into().unwrap());
    if u64::from(crc32c::crc32c(body)) != crc {
        return Err(DecodeError::BadCrc);
    }
    let kind = u16::from_le_bytes(bytes[4..6].try_into().unwrap());
    let flags = u16::from_le_bytes(bytes[6..8].try_into().unwrap());
    let request_id = u64::from_le_bytes(bytes[8..16].try_into().unwrap());
    let epoch = u64::from_le_bytes(bytes[16..24].try_into().unwrap());
    let ncols = u32::from_le_bytes(bytes[24..28].try_into().unwrap()) as usize;
    let nrows = u32::from_le_bytes(bytes[28..32].try_into().unwrap());
    let payload_len = u64::from_le_bytes(bytes[32..40].try_into().unwrap()) as usize;

    let dir_end = HEADER_LEN + ncols * 16;
    if dir_end > body.len() || HEADER_LEN + payload_len > body.len() {
        return Err(DecodeError::BadDirectory);
    }
    let payload = &body[dir_end..HEADER_LEN + payload_len];
    let mut cols = Vec::with_capacity(ncols);
    for i in 0..ncols {
        let at = HEADER_LEN + i * 16;
        let col_type = match bytes[at] {
            1 => ColType::U32,
            2 => ColType::U64,
            3 => ColType::F64,
            4 => ColType::Str,
            _ => return Err(DecodeError::BadDirectory),
        };
        let off = u64::from_le_bytes(bytes[at + 8..at + 16].try_into().unwrap()) as usize;
        if off > payload.len() {
            return Err(DecodeError::BadDirectory);
        }
        // Column extent = next column's offset (or payload end); views are
        // sliced generously — readers take nrows-bounded prefixes.
        let end = if i + 1 < ncols {
            let next_at = HEADER_LEN + (i + 1) * 16;
            u64::from_le_bytes(bytes[next_at + 8..next_at + 16].try_into().unwrap()) as usize
        } else {
            payload.len()
        };
        cols.push((
            col_type,
            payload.get(off..end).ok_or(DecodeError::BadDirectory)?,
        ));
    }
    Ok(FrameView {
        kind,
        flags,
        request_id,
        epoch,
        nrows,
        cols,
        payload,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip_all_column_types() {
        let strs = vec!["alpha".to_string(), String::new(), "γreek".to_string()];
        let frame = encode_frame(
            FrameKind::TopFunctions,
            FLAG_PARTIAL_TAIL,
            42,
            7,
            &[
                Col::U32(&[1, 2, 3]),
                Col::U64(&[10, 20, 30]),
                Col::F64(&[0.5, 1.5, 2.5]),
                Col::Str(&strs),
            ],
        );
        let view = decode_frame(&frame).unwrap();
        assert_eq!(view.kind, FrameKind::TopFunctions as u16);
        assert_eq!(view.flags, FLAG_PARTIAL_TAIL);
        assert_eq!(view.request_id, 42);
        assert_eq!(view.epoch, 7);
        assert_eq!(view.nrows, 3);
        assert_eq!(view.col_u32(0).unwrap(), vec![1, 2, 3]);
        assert_eq!(view.col_u64(1).unwrap(), vec![10, 20, 30]);
        assert_eq!(view.col_str(3).unwrap(), strs);
    }

    #[test]
    fn corrupt_frame_fails_crc() {
        let mut frame = encode_frame(FrameKind::Status, 0, 1, 0, &[Col::U32(&[9])]);
        let mid = frame.len() / 2;
        frame[mid] ^= 0xFF;
        assert_eq!(decode_frame(&frame).unwrap_err(), DecodeError::BadCrc);
    }
}

File: typescript2/pkg-playground/src/obs/bqf1.ts
/**
 * §9.3 BQF1: fixed little-endian columnar frame decoder.
 *
 * Mirrors the frozen Rust encoder in `crates/bex_query/src/bqf1.rs`:
 *
 * ```text
 * [40 B header]
 *   magic        4B  "BQF1"
 *   kind         u16 @4
 *   flags        u16 @6   (bit0 lod_degraded, bit1 partial_tail, bit2 more_lanes)
 *   request_id   u64 @8
 *   epoch        u64 @16
 *   ncols        u32 @24
 *   nrows        u32 @28
 *   payload_len  u64 @32  (directory + columns, excluding trailer)
 * [column directory] ncols x 16 B @40: (col_type u8, pad[7], offset u64
 *   relative to PAYLOAD START = byte 40 + ncols*16)
 * [column payloads]  each 8-byte aligned
 *   1 u32: nrows x 4 B    2 u64: nrows x 8 B    3 f64: nrows x 8 B
 *   4 str: (nrows+1) x u32 offsets, pad8, utf8 bytes
 * [8 B trailer] crc32c of everything before it, stored as u64 LE
 * ```
 *
 * Numeric columns decode to zero-copy TypedArray views over the frame's
 * ArrayBuffer (payload start = 40 + ncols*16 is 8-aligned, and per-column
 * offsets are 8-aligned, so views are always aligned when the frame starts
 * at buffer offset 0 — which a WebSocket binary message guarantees).
 * Strings decode eagerly to string[].
 */

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

export const BQF1_HEADER_LEN = 40;

export const FLAG_LOD_DEGRADED = 1 << 0;
export const FLAG_PARTIAL_TAIL = 1 << 1;
export const FLAG_MORE_LANES = 1 << 2;

/** Frame kinds (u16 in the header). */
export const FrameKind = {
  BqlTable: 9,
  LeftHeavy: 4,
  LiveTotals: 7,
  RecentCalls: 8,
  RunMeta: 2,
  RunsList: 1,
  Status: 6,
  Timeline: 3,
  TopFunctions: 5,
} as const;
export type FrameKind = (typeof FrameKind)[keyof typeof FrameKind];

/** LeftHeavy `function` sentinel: a synthetic "N smaller" fold row. */
export const FOLD_ROW_FUNCTION = 0xffff_ffff;

// ---------------------------------------------------------------------------
// crc32c (Castagnoli, reflected, poly 0x82F63B78) — matches the Rust
// implementation in bex_events::prof::cct::crc32c.
// ---------------------------------------------------------------------------

const CRC32C_TABLE: Uint32Array = (() => {
  const table = new Uint32Array(256);
  for (let i = 0; i < 256; i += 1) {
    let crc = i;
    for (let bit = 0; bit < 8; bit += 1) {
      crc = crc & 1 ? (crc >>> 1) ^ 0x82f63b78 : crc >>> 1;
    }
    table[i] = crc >>> 0;
  }
  return table;
})();

/** One-shot CRC32C (init 0xFFFFFFFF, final xor 0xFFFFFFFF). */
export function crc32c(bytes: Uint8Array): number {
  let crc = 0xffffffff;
  for (let i = 0; i < bytes.length; i += 1) {
    crc = (crc >>> 8) ^ CRC32C_TABLE[(crc ^ bytes[i]!) & 0xff]!;
  }
  return (crc ^ 0xffffffff) >>> 0;
}

// ---------------------------------------------------------------------------
// Decoded frame
// ---------------------------------------------------------------------------

export type BqfColumn =
  | { type: 'u32'; data: Uint32Array }
  | { type: 'u64'; data: BigUint64Array }
  | { type: 'f64'; data: Float64Array }
  | { type: 'str'; data: string[] };

export interface BqfFrame {
  kind: number;
  flags: number;
  /** request_id echoed from the query/sub JSON `id` (u64; ids are small). */
  requestId: number;
  /** Data generation the frame was computed at. */
  epoch: number;
  nrows: number;
  cols: BqfColumn[];
}

export class BqfDecodeError extends Error {}

const utf8 = new TextDecoder('utf-8');

/**
 * Decode one BQF1 frame. Throws `BqfDecodeError` on truncation, bad magic,
 * CRC mismatch, or a malformed directory.
 */
export function decodeFrame(buf: ArrayBuffer): BqfFrame {
  const bytes = new Uint8Array(buf);
  if (bytes.length < BQF1_HEADER_LEN + 8) {
    throw new BqfDecodeError('BQF1 frame too short');
  }
  // magic "BQF1"
  if (
    bytes[0] !== 0x42 ||
    bytes[1] !== 0x51 ||
    bytes[2] !== 0x46 ||
    bytes[3] !== 0x31
  ) {
    throw new BqfDecodeError('BQF1 bad magic');
  }

  const view = new DataView(buf);
  const body = bytes.subarray(0, bytes.length - 8);
  const storedCrc = view.getBigUint64(bytes.length - 8, true);
  if (BigInt(crc32c(body)) !== storedCrc) {
    throw new BqfDecodeError('BQF1 CRC mismatch');
  }

  const kind = view.getUint16(4, true);
  const flags = view.getUint16(6, true);
  const requestId = Number(view.getBigUint64(8, true));
  const epoch = Number(view.getBigUint64(16, true));
  const ncols = view.getUint32(24, true);
  const nrows = view.getUint32(28, true);
  const payloadLen = Number(view.getBigUint64(32, true));

  const dirEnd = BQF1_HEADER_LEN + ncols * 16;
  if (dirEnd > body.length || BQF1_HEADER_LEN + payloadLen > body.length) {
    throw new BqfDecodeError('BQF1 bad directory');
  }
  const payloadStart = dirEnd; // absolute byte offset of payload start
  const payloadEnd = BQF1_HEADER_LEN + payloadLen;

  const colOffset = (i: number): number =>
    Number(view.getBigUint64(BQF1_HEADER_LEN + i * 16 + 8, true));

  const cols: BqfColumn[] = [];
  for (let i = 0; i < ncols; i += 1) {
    const colType = bytes[BQF1_HEADER_LEN + i * 16]!;
    const off = colOffset(i);
    const at = payloadStart + off;
    // Column extent = next column's offset (or payload end); readers take
    // nrows-bounded prefixes, mirroring the Rust decoder.
    const end = i + 1 < ncols ? payloadStart + colOffset(i + 1) : payloadEnd;
    if (at > payloadEnd || end > payloadEnd || at > end) {
      throw new BqfDecodeError('BQF1 bad column offset');
    }
    switch (colType) {
      case 1: {
        if (at + nrows * 4 > end)
          throw new BqfDecodeError('BQF1 u32 column truncated');
        cols.push({ data: new Uint32Array(buf, at, nrows), type: 'u32' });
        break;
      }
      case 2: {
        if (at + nrows * 8 > end)
          throw new BqfDecodeError('BQF1 u64 column truncated');
        cols.push({ data: new BigUint64Array(buf, at, nrows), type: 'u64' });
        break;
      }
      case 3: {
        if (at + nrows * 8 > end)
          throw new BqfDecodeError('BQF1 f64 column truncated');
        cols.push({ data: new Float64Array(buf, at, nrows), type: 'f64' });
        break;
      }
      case 4: {
        cols.push({ data: decodeStrColumn(buf, at, end, nrows), type: 'str' });
        break;
      }
      default:
        throw new BqfDecodeError(`BQF1 unknown column type ${colType}`);
    }
  }

  return { cols, epoch, flags, kind, nrows, requestId };
}

function decodeStrColumn(
  buf: ArrayBuffer,
  at: number,
  end: number,
  nrows: number,
): string[] {
  const offsetsLen = (nrows + 1) * 4;
  if (at + offsetsLen > end) {
    throw new BqfDecodeError('BQF1 str offsets truncated');
  }
  // `at` is 8-aligned, so a u32 view is aligned.
  const offsets = new Uint32Array(buf, at, nrows + 1);
  // The utf8 blob starts at the 8-aligned boundary after the offsets.
  let blobAt = at + offsetsLen;
  blobAt += (8 - (blobAt % 8)) % 8;
  const blob = new Uint8Array(buf, blobAt, end - blobAt);
  const out: string[] = [];
  for (let i = 0; i < nrows; i += 1) {
    const from = offsets[i]!;
    const to = offsets[i + 1]!;
    if (from > to || to > blob.length) {
      throw new BqfDecodeError('BQF1 str extent out of range');
    }
    out.push(from === to ? '' : utf8.decode(blob.subarray(from, to)));
  }
  return out;
}

// ---------------------------------------------------------------------------
// Typed per-kind helpers
// ---------------------------------------------------------------------------

interface ColDataByType {
  u32: Uint32Array;
  u64: BigUint64Array;
  f64: Float64Array;
  str: string[];
}

function expectCol<T extends keyof ColDataByType>(
  frame: BqfFrame,
  index: number,
  type: T,
  kindName: string,
): ColDataByType[T] {
  const col = frame.cols[index];
  if (!col || col.type !== type) {
    throw new BqfDecodeError(
      `BQF1 ${kindName}: expected col ${index} of type ${type}, got ${col?.type ?? 'missing'}`,
    );
  }
  return col.data as ColDataByType[T];
}

function expectKind(frame: BqfFrame, kind: FrameKind, kindName: string): void {
  if (frame.kind !== kind) {
    throw new BqfDecodeError(
      `BQF1: expected ${kindName} (kind ${kind}), got kind ${frame.kind}`,
    );
  }
}

/** Convert a u64 column to JS numbers (values fit: ns within runs, ms epochs). */
export function u64ToNumbers(data: BigUint64Array): Float64Array {
  const out = new Float64Array(data.length);
  for (let i = 0; i < data.length; i += 1) out[i] = Number(data[i]!);
  return out;
}

export interface RunsListColumns {
  runKey: string[];
  boundaryId: string[];
  target: string[];
  source: string[];
  status: string[];
  revision: string[];
  createdMs: Float64Array;
  completedMs: Float64Array;
  hasSnapshot: Uint32Array;
}

export function asRunsList(frame: BqfFrame): RunsListColumns {
  expectKind(frame, FrameKind.RunsList, 'RunsList');
  return {
    boundaryId: expectCol(frame, 1, 'str', 'RunsList'),
    completedMs: u64ToNumbers(expectCol(frame, 7, 'u64', 'RunsList')),
    createdMs: u64ToNumbers(expectCol(frame, 6, 'u64', 'RunsList')),
    hasSnapshot: expectCol(frame, 8, 'u32', 'RunsList'),
    revision: expectCol(frame, 5, 'str', 'RunsList'),
    runKey: expectCol(frame, 0, 'str', 'RunsList'),
    source: expectCol(frame, 3, 'str', 'RunsList'),
    status: expectCol(frame, 4, 'str', 'RunsList'),
    target: expectCol(frame, 2, 'str', 'RunsList'),
  };
}

export interface RunMetaColumns {
  functionId: Uint32Array;
  fqn: string[];
}

export function asRunMeta(frame: BqfFrame): RunMetaColumns {
  expectKind(frame, FrameKind.RunMeta, 'RunMeta');
  return {
    fqn: expectCol(frame, 1, 'str', 'RunMeta'),
    functionId: expectCol(frame, 0, 'u32', 'RunMeta'),
  };
}

export interface TimelineColumns {
  /** Raw u64 thread ids (may exceed 2^53; use the raw view for identity). */
  thread: BigUint64Array;
  firstTsNs: Float64Array;
  lastTsNs: Float64Array;
  busyNs: Float64Array;
  awaitNs: Float64Array;
  dominantFunction: Uint32Array;
  errors: Float64Array;
}

export function asTimeline(frame: BqfFrame): TimelineColumns {
  expectKind(frame, FrameKind.Timeline, 'Timeline');
  return {
    awaitNs: u64ToNumbers(expectCol(frame, 4, 'u64', 'Timeline')),
    busyNs: u64ToNumbers(expectCol(frame, 3, 'u64', 'Timeline')),
    dominantFunction: expectCol(frame, 5, 'u32', 'Timeline'),
    errors: u64ToNumbers(expectCol(frame, 6, 'u64', 'Timeline')),
    firstTsNs: u64ToNumbers(expectCol(frame, 1, 'u64', 'Timeline')),
    lastTsNs: u64ToNumbers(expectCol(frame, 2, 'u64', 'Timeline')),
    thread: expectCol(frame, 0, 'u64', 'Timeline'),
  };
}

export interface LeftHeavyColumns {
  depth: Uint32Array;
  /** 0xFFFFFFFF (`FOLD_ROW_FUNCTION`) marks a synthetic "smaller" fold row. */
  functionId: Uint32Array;
  totalNs: Float64Array;
  selfNs: Float64Array;
  enters: Float64Array;
  errors: Float64Array;
  foldedCount: Uint32Array;
}

export function asLeftHeavy(frame: BqfFrame): LeftHeavyColumns {
  expectKind(frame, FrameKind.LeftHeavy, 'LeftHeavy');
  return {
    depth: expectCol(frame, 0, 'u32', 'LeftHeavy'),
    enters: u64ToNumbers(expectCol(frame, 4, 'u64', 'LeftHeavy')),
    errors: u64ToNumbers(expectCol(frame, 5, 'u64', 'LeftHeavy')),
    foldedCount: expectCol(frame, 6, 'u32', 'LeftHeavy'),
    functionId: expectCol(frame, 1, 'u32', 'LeftHeavy'),
    selfNs: u64ToNumbers(expectCol(frame, 3, 'u64', 'LeftHeavy')),
    totalNs: u64ToNumbers(expectCol(frame, 2, 'u64', 'LeftHeavy')),
  };
}

export interface TopFunctionsColumns {
  functionId: Uint32Array;
  calls: Float64Array;
  totalNs: Float64Array;
  selfNs: Float64Array;
  errors: Float64Array;
}

export function asTopFunctions(frame: BqfFrame): TopFunctionsColumns {
  expectKind(frame, FrameKind.TopFunctions, 'TopFunctions');
  return {
    calls: u64ToNumbers(expectCol(frame, 1, 'u64', 'TopFunctions')),
    errors: u64ToNumbers(expectCol(frame, 4, 'u64', 'TopFunctions')),
    functionId: expectCol(frame, 0, 'u32', 'TopFunctions'),
    selfNs: u64ToNumbers(expectCol(frame, 3, 'u64', 'TopFunctions')),
    totalNs: u64ToNumbers(expectCol(frame, 2, 'u64', 'TopFunctions')),
  };
}

export interface RecentCallsColumns {
  /** Raw u64 `(partition << 32) | thread_idx` ids; use as identity only. */
  thread: BigUint64Array;
  /** Per-thread call ids; unique key is `(thread, callId)`. */
  callId: BigUint64Array;
  parentCallId: BigUint64Array;
  functionId: Uint32Array;
  /**
   * Raw u64 nanosecond timestamps. Kept as BigUint64Array so callers can
   * subtract a baseline in BigInt space before converting to Number (epoch-ns
   * values exceed 2^53; deltas within a run do not).
   */
  startNs: BigUint64Array;
  endNs: BigUint64Array;
  /** FunctionEndStatus: 0 ok, 1 errored, 2 cancelled, 3 exited. */
  status: Uint32Array;
}

/** §9.4 exact-recency tier: completed calls from the live engine's rings. */
export function asRecentCalls(frame: BqfFrame): RecentCallsColumns {
  expectKind(frame, FrameKind.RecentCalls, 'RecentCalls');
  return {
    callId: expectCol(frame, 1, 'u64', 'RecentCalls'),
    endNs: expectCol(frame, 5, 'u64', 'RecentCalls'),
    functionId: expectCol(frame, 3, 'u32', 'RecentCalls'),
    parentCallId: expectCol(frame, 2, 'u64', 'RecentCalls'),
    startNs: expectCol(frame, 4, 'u64', 'RecentCalls'),
    status: expectCol(frame, 6, 'u32', 'RecentCalls'),
    thread: expectCol(frame, 0, 'u64', 'RecentCalls'),
  };
}

export interface StatusColumns {
  code: Uint32Array;
  message: string[];
}

/**
 * Generic BQL result (kind 9): free-form data columns + a final Str meta
 * column. Frame row 0 is a meta row (sentinels; the meta col's row 0 holds
 * `{"columns":[{name,type}...],"rows":N,"footer":{...}}`); data row i is
 * frame row i+1 — so an empty result still ships its footer.
 */
export interface BqlTableResult {
  columns: Array<{ name: string; values: Array<string | number> }>;
  rows: number;
  footer: {
    sealed: boolean;
    torn: boolean;
    degraded: string[];
  };
}

export function asBqlTable(frame: BqfFrame): BqlTableResult {
  expectKind(frame, FrameKind.BqlTable, 'BqlTable');
  const metaCol = frame.cols[frame.cols.length - 1];
  if (!metaCol || metaCol.type !== 'str') {
    throw new BqfDecodeError('BQF1 BqlTable: missing meta column');
  }
  const metaRaw = metaCol.data[0] ?? '';
  let meta: {
    columns?: Array<{ name?: string; type?: string }>;
    rows?: number;
    footer?: { sealed?: boolean; torn?: boolean; degraded?: string[] };
  };
  try {
    meta = JSON.parse(metaRaw) as typeof meta;
  } catch {
    throw new BqfDecodeError('BQF1 BqlTable: bad meta JSON');
  }
  const described = meta.columns ?? [];
  const dataRows = Math.max(0, frame.nrows - 1);
  const columns: BqlTableResult['columns'] = [];
  for (let i = 0; i < described.length && i < frame.cols.length - 1; i += 1) {
    const col = frame.cols[i]!;
    const values: Array<string | number> = [];
    for (let row = 1; row <= dataRows; row += 1) {
      switch (col.type) {
        case 'u32':
          values.push(col.data[row] ?? 0);
          break;
        case 'u64':
          values.push(Number(col.data[row] ?? 0n));
          break;
        case 'f64':
          values.push(col.data[row] ?? 0);
          break;
        case 'str':
          values.push(col.data[row] ?? '');
          break;
      }
    }
    columns.push({ name: described[i]?.name ?? `col${i}`, values });
  }
  return {
    columns,
    footer: {
      degraded: meta.footer?.degraded ?? [],
      sealed: meta.footer?.sealed ?? true,
      torn: meta.footer?.torn ?? false,
    },
    rows: dataRows,
  };
}

export function asStatus(frame: BqfFrame): StatusColumns {
  expectKind(frame, FrameKind.Status, 'Status');
  return {
    code: expectCol(frame, 0, 'u32', 'Status'),
    message: expectCol(frame, 1, 'str', 'Status'),
  };
}
